File: stml_cpp/stml.cpp
#include "stml.h"
#include "lexer/lexer.h"
#include "parser/parser.h"
<<<<<<< Updated upstream
=======
<<<<<<< Updated upstream
>>>>>>> Stashed changes
#include <fstream>
#include <sstream>
#include <utility>

namespace stml {

// =========================================================================
// High-level API
// =========================================================================

LoadResult loads(const std::string& text) {
    Lexer lexer;
    std::vector<Token> tokens = lexer.tokenize(text);
    std::vector<Warning> warnings = lexer.warnings();

    Parser parser;
    AstNode ast = parser.parse(tokens);

    // Merge lexer warnings with parser warnings
    const auto& p_warnings = parser.warnings();
    warnings.insert(warnings.end(), p_warnings.begin(), p_warnings.end());

    return LoadResult(std::move(ast), std::move(warnings));
}

LoadResult load(const std::string& filepath) {
    std::ifstream file(filepath, std::ios::binary);
    if (!file.is_open()) {
        throw ParseError(0, 0, "Cannot open file: " + filepath);
    }
<<<<<<< Updated upstream
    std::ostringstream ss;
    ss << file.rdbuf();
    return loads(ss.str());
=======
=======
#include "serializer/serializer.h"
#include <fstream>
#include <sstream>

namespace stml {

AstList loads(const std::string& input) {
    Lexer lexer;
    auto tokens = lexer.tokenize(input);

    Parser parser;
    return parser.parse(tokens);
>>>>>>> Stashed changes
}

AstList load(const std::string& filepath) {
    std::ifstream file(filepath);
    if (!file.is_open()) {
        throw ParseError(0, 0, "Cannot open file: " + filepath);
    }

>>>>>>> Stashed changes
    std::ostringstream ss;
    ss << file.rdbuf();
    return loads(ss.str());
}

<<<<<<< Updated upstream
// =========================================================================
// Low-level / debugging API
// =========================================================================

std::pair<std::vector<Token>, std::vector<Warning>>
tokenize(const std::string& text) {
    Lexer lexer;
    std::vector<Token> tokens = lexer.tokenize(text);
    std::vector<Warning> warnings = lexer.warnings();
    return {std::move(tokens), std::move(warnings)};
}

std::pair<AstNode, std::vector<Warning>>
parse(const std::vector<Token>& tokens) {
    Parser parser;
    AstNode ast = parser.parse(tokens);
    std::vector<Warning> warnings = parser.warnings();
    return {std::move(ast), std::move(warnings)};
<<<<<<< Updated upstream
=======
=======
std::string dumps(const AstList& docs, int indent) {
    return docs_to_stml(docs, indent);
}

std::string to_json(const AstList& docs, int indent) {
    return docs_to_json(docs, indent);
>>>>>>> Stashed changes
>>>>>>> Stashed changes
}

} // namespace stml

File: stml_java/src/main/java/stml/STML.java
package stml;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

public class STML {

    public static class LoadResult {
        public final AstNode ast;
        public final List<Warning> warnings;
        public LoadResult(AstNode ast, List<Warning> warnings) {
            this.ast = ast;
            this.warnings = warnings;
        }
    }

    public static LoadResult loads(String text) {
        STMLLexer lexer = new STMLLexer(text);
        STMLLexer.TokenizeResult lexResult = lexer.tokenize();
        List<Token> tokens = lexResult.tokens;
        List<Warning> lexWarnings = lexResult.warnings;

        LineTreeBuilder treeBuilder = new LineTreeBuilder();
        treeBuilder.process(tokens);

        AstBuilder astBuilder = new AstBuilder();
        List<AstNode> docsList = new ArrayList<>();
        for (List<Line> docLines : treeBuilder.documents()) {
            docsList.add(astBuilder.build(docLines));
        }

        List<Warning> allWarnings = new ArrayList<>();
        allWarnings.addAll(lexWarnings);
        allWarnings.addAll(treeBuilder.warnings());
        allWarnings.addAll(astBuilder.warnings());

        List<AstNode.Pair> wrapper = new ArrayList<>();
        wrapper.add(new AstNode.Pair("docs", new AstNode(docsList)));
        return new LoadResult(new AstNode(wrapper), allWarnings);
    }

    public static LoadResult load(String filename) throws IOException {
        String text = Files.readString(Path.of(filename));
        return loads(text);
    }

    public static String dumps(AstNode ast) {
        return Serializer.dumps(ast);
    }

    public static String toJson(AstNode ast) {
        return Serializer.toJson(ast);
    }

    public static STMLLexer.TokenizeResult tokenize(String text) {
        STMLLexer lexer = new STMLLexer(text);
        return lexer.tokenize();
    }

    public static class ParseResult {
        public final List<AstNode> astList;
        public final List<Warning> warnings;
        public ParseResult(List<AstNode> astList, List<Warning> warnings) {
            this.astList = astList;
            this.warnings = warnings;
        }
    }

    public static ParseResult parse(List<Token> tokens) {
        LineTreeBuilder treeBuilder = new LineTreeBuilder();
        treeBuilder.process(tokens);

        AstBuilder astBuilder = new AstBuilder();
        List<AstNode> docsList = new ArrayList<>();
        for (List<Line> docLines : treeBuilder.documents()) {
            docsList.add(astBuilder.build(docLines));
        }

        List<Warning> allWarnings = new ArrayList<>();
        allWarnings.addAll(treeBuilder.warnings());
        allWarnings.addAll(astBuilder.warnings());

        return new ParseResult(docsList, allWarnings);
    }
}